File: poa-search-api/src/main/java/nl/rabobank/service/poa/search/endpoint/DetailedPowerOfAttorneyController.java
package nl.rabobank.service.poa.search.endpoint;


import io.swagger.annotations.Api;
import io.swagger.annotations.ApiOperation;
import nl.rabobank.service.poa.search.detailedpowerofattorney.DetailedPowerOfAttorney;
import nl.rabobank.service.poa.search.detailedpowerofattorney.DetailedPowerOfAttorneys;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import javax.annotation.Nonnull;

@Api(value = "detailed-power-of-attorney", description = "Detailed Power of Attorney for a client")
@RestController
@RequestMapping(value = "/v1/detailed-power-of-attorney", produces = MediaType.APPLICATION_JSON_VALUE)
public interface DetailedPowerOfAttorneyController {

    @ApiOperation(value = "Retrieves detailed Power of Attorney information for a client")
    @GetMapping
    DetailedPowerOfAttorneys getDetailedPowerOfAttorneyForClient(@Nonnull @RequestParam(value = "client-name") final String clientName);

    @ApiOperation(value = "Retrieves detailed Power of Attorney information for a power of attorney by id")
    @RequestMapping(value = "/{power-of-attorney-id}", method = RequestMethod.GET)
    DetailedPowerOfAttorney getDetailedPowerOfAttorneyByPowerOfAttorneyId(@Nonnull @PathVariable(value = "power-of-attorney-id") final String powerOfAttorneyId);
}

File: poa-search-server/src/main/kotlin/nl/rabobank/service/poa/search/detailedpowerofattorney/DetailedPowerOfAttorneyController.kt
package nl.rabobank.service.poa.search.detailedpowerofattorney

import nl.rabobank.service.poa.search.detailedpowerofattorney.model.DetailedPowerOfAttorneys
import nl.rabobank.service.poa.search.endpoint.DetailedPowerOfAttorneyController
import nl.rabobank.service.poa.search.exception.UnauthorizedResourceException
import nl.rabobank.service.poa.search.util.isSecurityEnabled
import org.springframework.core.env.Environment
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.annotation.RestController
import nl.rabobank.service.poa.search.detailedpowerofattorney.DetailedPowerOfAttorneys as ExternalDetailedPowerOfAttorneys
import nl.rabobank.service.poa.search.detailedpowerofattorney.DetailedPowerOfAttorney as ExternalDetailedPowerOfAttorney
import nl.rabobank.service.poa.search.detailedpowerofattorney.model.DetailedPowerOfAttorney

@RestController
class DetailedPowerOfAttorneyController(
        private val detailedPowerOfAttorneyService: DetailedPowerOfAttorneyService,
        private val environment: Environment) : DetailedPowerOfAttorneyController {

    override fun getDetailedPowerOfAttorneyForClient(clientName: String): ExternalDetailedPowerOfAttorneys {
        validateAuthorization(clientName = clientName)

        return DetailedPowerOfAttorneys(detailedPowerOfAttorneyService.getDetailedPowerOfAttorneyByName(clientName)).expose()
    }

    override fun getDetailedPowerOfAttorneyByPowerOfAttorneyId(powerOfAttorneyId: String): ExternalDetailedPowerOfAttorney {
        val detailedPowerOfAttorney = detailedPowerOfAttorneyService.getDetailedPowerOfAttorneyByPowerOfAttorneyId(powerOfAttorneyId)

        validateAuthorization(detailedPowerOfAttorney)

        return detailedPowerOfAttorney.expose()
    }

    private fun validateAuthorization(clientName: String) {
        if (environment.isSecurityEnabled() &&
                clientName != SecurityContextHolder.getContext().authentication.name)
            throw UnauthorizedResourceException(SecurityContextHolder.getContext().authentication.name, clientName)
    }

    private fun validateAuthorization(detailedPowerOfAttorney: DetailedPowerOfAttorney) {
        if (environment.isSecurityEnabled() &&
                detailedPowerOfAttorney.grantee != SecurityContextHolder.getContext().authentication.name
                && detailedPowerOfAttorney.grantor != SecurityContextHolder.getContext().authentication.name)
            throw UnauthorizedResourceException(SecurityContextHolder.getContext().authentication.name, detailedPowerOfAttorney.grantor)
    }
}